user/utest/misbehave.c

```diff
@@ -24,3 +27,44 @@ int test_misbehave() {
 
   return 0;
 }
+
+static inline void exec_cp0_instr(void) {
+  int value;
+  asm volatile("mfc0 %0, $12, 0" : "=r"(value));
+}
+
+static inline void exec_reserved_instr(void) {
+  int value;
+  asm volatile("mfc2 %0, $12, 0" : "=r"(value));
```

Comment:
Some MIPS CPUs implement CP2 as mentioned on wikipedia:

> For example, in the PlayStation video game console, CP2 is the Geometry Transformation Engine (GTE), which accelerates the processing of geometry in 3D computer graphics.

Reserved instructions are according to my knowledge all privileged instructions such as `eret`, `tlbp` and so on. Refer to **Table 3.23 Privileged Instructions** from *MIPS® Architecture For Programmers Volume II-A: The MIPS32® Instruction Set*.

Reply:
Suprisingly all of the instructions from Table 3.23 cause Coprocessor Unusable exception. At least on ours Qemu setup.

Reply:
Right! Another idea – refer to chapter A.2 of same document. Choose encoding that has `*` in it. According to *Table A.1 Symbols Used in the Instruction Encoding Tables*:

> Operation or field codes marked with this symbol are reserved for future use. Executing such an instruction must cause a Reserved Instruction Exception.